wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivityManager.java

```diff
@@ -167,12 +171,82 @@ protected long getIdleTimeout(String wsId) {
       delayParameterName = "che.workspace.activity_check_scheduler_period_s")
   private void invalidate() {
     try {
-      activityDao.findExpired(System.currentTimeMillis()).forEach(this::stopExpired);
+      stopAllExpired();
+      checkValidExpiration();
     } catch (ServerException e) {
       LOG.error(e.getLocalizedMessage(), e);
     }
   }
 
+  private void stopAllExpired() throws ServerException {
+    activityDao.findExpired(System.currentTimeMillis()).forEach(this::stopExpired);
+  }
+
+  private void checkValidExpiration() throws ServerException {
+    for (String runningWsId : workspaceRuntimes.getRunning()) {
+      WorkspaceActivity activity = activityDao.findActivity(runningWsId);
+      if (activity == null) {
+        LOG.warn(
+            "Found a running workspace {} without any activity record. This shouldn't really"
+                + " happen but is being rectified by adding a new activity record for it.",
+            runningWsId);
+
+        try {
+          Workspace workspace = workspaceManager.getWorkspace(runningWsId);
+          long createdTime =
+              Long.parseLong(workspace.getAttributes().get(Constants.CREATED_ATTRIBUTE_NAME));
+
+          activity = new WorkspaceActivity();
+          activity.setWorkspaceId(runningWsId);
+          activity.setCreated(createdTime);
+          activity.setStatus(WorkspaceStatus.RUNNING);
+          activity.setLastRunning(System.currentTimeMillis());
+
+          long idleTimeout = getIdleTimeout(runningWsId);
+          if (idleTimeout > 0) {
+            // only set the expiration if it is used...
+            activity.setExpiration(System.currentTimeMillis() + idleTimeout);
+          }
+
+          activityDao.createActivity(activity);
+        } catch (NotFoundException e) {
+          LOG.error(
+              "Detected a running workspace {} but could not find its record.", runningWsId, e);
+        } catch (ConflictException e) {
+          LOG.debug(
+              "Activity record created while we were trying to rectify its absence for a"
+                  + " running workspace {}.",
+              runningWsId);
+        }
+      } else {
+        long idleTimeout = getIdleTimeout(runningWsId);
+        if (idleTimeout == 0) {
+          // expiry not used at all...
+          continue;
+        }
+
+        long expectedExpiryTime =
+            valueOrDefault(activity.getLastRunning(), System.currentTimeMillis()) + idleTimeout;
+        long actualExpiryTime = valueOrDefault(activity.getExpiration(), 0);
+
+        if (actualExpiryTime != expectedExpiryTime) {
```

Comment:
Sorry, I don't understand this check. 

> expectedExpiryTime = valueOrDefault(activity.getLastRunning(), System.currentTimeMillis()) + idleTimeout;

I think it does not make sense to check  `System.currentTimeMillis() + idleTimeout` if `getLastRunning` is absent. 
What is the probability that expires time was set in the same millisecond?

Also, I think that that `actualExpiryTime` should not equal `activity.getLastRunning() + idleTimeout`. Because `getLastRunning` is timestamp when workspace was started, right?
If yes, then a user may use this workspace and expiration time will be updated to like `activity.getLastRunning() + 2*idleTimeout` and it's a valid case.

Reply:
Well, you're right that I missed the fact that the expiry period is updated remotely from the clients on user activity, that's really stupid of me.

On the other hand, the code as is atm deals with the case where the expiration timestamp is not set when it should have been. For that, we need a value to set the expiration period to, and IMHO current time + the timeout is a reasonable value to use. It will be updated if there is any user activity on the workspace. This will also happen when the schedule exactly coincides with the workspace transitioning to RUNNING, but no harm should be done here - the two expiration times will be almost identical depending on which transaction commits last.

As for the `!=` - I agree that rarely we are going to see that the first time the expiry is set, it is going to be in the exact millisecond. The reason why I used `!=` here is to be able to settle on a value eventually.

It is obviously wrong though, because it doesn't account for the updates due to the user activity - I should have used `<` instead - the expiry time should never be less than the time the workspace started + the timeout.

If it is, then the expiry must not have been set when the workspace started, which is wrong.

So IMHO, it is enough to change the code to use `actualExpiryTime < expectedExpiryTime` and it should work correctly. Also I should change the log level of the subsequent message to at least INFO or maybe even warning, because it really is an error condition that we're correcting there.